### DISCUSSION TITLE
graphviz fixes

### DISCUSSION THREAD

**Discussion by revelator** [msys2/MINGW-packages#8096] [Show and tell]
After my initial commit with some fixes for the graphviz package i have done some further work.
gdiplus plugin working with mingw-w64 now.
gvedit builds and works now.
swig plugins fixed (go and python plugins built by default).
smyrna disabled in PKGBUILD since it is crash prone as hell.

if you can use my fixes feel free to add them.

[mingw-w64-graphviz-fixups.zip](https://github.com/msys2/MINGW-packages/files/6107722/mingw-w64-graphviz-fixups.zip)

**Comment by revelator** [comment]
![gvedit](https://user-images.githubusercontent.com/1930128/110456998-78996500-80ca-11eb-9a45-927fb808c56d.jpg)

**Comment by mingwandroid** [comment]
Hi @revelator, I know you said you cannot make PRs, can you explain why? It's great that you're doing this work but it'd be greater if we could credit you and not have to do the last bits :-)

**Comment by revelator** [comment]
Well i setup a fork sadly when i tried to make a pull request from it they all ended up under the same initial commit breaking everything so id rather not try that again untill i know why this happened.

Btw noticed that gvedit could not find atrributes.txt ? looks like it has not been updated for ages since the txt file is now named attrs.txt i corrected it in my next patch.

**Reply by Biswa96** [reply]
The fear of screwing up everything, I understand 😌

**Reply by revelator** [reply]
xD indeed, id rather not break your ci.

**Reply by mingwandroid** [reply]
You cannot "break our CI" with a PR though, for that someone would need to merge a broken commit.

**Reply by revelator** [reply]
hehe id rather be sure than sorry :) still i need to figure out how to seperate pull requests so that they not all end up under the same item.

**Reply by naveen521kk** [reply]
I would suggest you use [github-cli](https://packages.msys2.org/base/mingw-w64-github-cli) :).

**Reply by revelator** [reply]
Ok i was doing pull requests directly using githubs internal tools but they made a mess of things :S ill try github-cli though i might need some help setting it up.

**Reply by revelator** [reply]
Oh its a go tool ?!? and seems i need to update go aswell... --buildmode=pie is unsupported when i try building it.

**Reply by naveen521kk** [reply]
You can simply use pre-build binaries. I don't see the point it building it when you want to use it.

**Reply by revelator** [reply]
well the thing is i use an offshot of your package with sjlj exceptions instead of dwarf and seh so to keep things neat id prefer having it built natively. The build is not a problem what i might need help with is getting it setup to create pull requests here :)

**Reply by naveen521kk** [reply]
you can use plain git also if you know what you are doing :)

**Reply by revelator** [reply]
ok it is up and device activated.
now i just need to figure out the commands for making a pull request ?.

**Reply by revelator** [reply]
well it kinda works untill it does not... cannot use it via bash as it croacks on windows paths.
had to read up a bit on using tortoisegit to do it and i just made a pull request for this though the tag is wrongly named 0, 013-further-fixes.patch etc.

**Reply by Alexpux** [reply]
@revelator Why not then use Msys2 git + tig? It simple to use

**Reply by revelator** [reply]
Hmm ci build seems to think one of the files has wrong checksums so the build fails sigh...

**Reply by naveen521kk** [reply]
I think the simplest suggestion I can give now is going to https://github.com/msys2/MINGW-packages/edit/master/mingw-w64-graphviz/PKGBUILD edit it say add a space and commit it. it will create a branch called `patch-1` or something with the current master in yours fork. Checkout the branch locally do the changes commit it and push it then make a PR.

**Reply by revelator** [reply]
"005-plugin-exports.patch ... FAILED"
which is strange as i did do an updpkgsums before pushing it to my work branch (scratches head ''').
i cannot believe it has to be this complicated to do a simple pull request.

and it builds locally with no such errors so im at a loss.

**Reply by revelator** [reply]
oh spew for some reason the corrected PKGBUILD newer ended up in my work branch, had to push it again crossing fingers this time.

**Reply by revelator** [reply]
well atleast it started building it now though i noticed i forgot to add qt5 as a dependency and you also need to add -fcommon to flags or you get mutiple definition errors quite early.

**Reply by revelator** [reply]
corrected the missing dependency and added the -fcommon flag to the existing flags in the PKGBUILD, ci now builds it with no errors.

**Comment by revelator** [comment]
updated patch with a fix for the wrong attributes.txt 
[013-further-fixes.zip](https://github.com/msys2/MINGW-packages/files/6108636/013-further-fixes.zip)

now to make ocaml and lua plugins work and i think we can call it a day, lua kinda gets picked up but since we have 2 versions the autotools get confused and try to link to the wrong version. Ocaml might be more than i can handle alone since i know little to nothing besides it being a bytecode compiler about it.

**Comment by Biswa96** [comment]
The patch file changes a lot. Have you asked this issue in upstream?

**Reply by revelator** [reply]
no and it might be problematic using it on say nix as i changed the .so files to .dll so that they work on windows.
patch can be used for mingw-w64 windows builds though but might point out that it is optional.

ci build just completed with no errors :) give it a spin and let me know if there are any problems.

**Reply by revelator** [reply]
someone more skilled in autotools could probably correct it so that it works for both arch's but as the package is ment for us im a bit partial as to the need.

**Reply by revelator** [reply]
Maybe an idea but i could remove the huge block of changes for the swig plugins, i actually dont know if they were ever intended for a windows build so we might save some space that way. gdiplus is windows only so that one seems to fit the bill, and gvedit is crossplatform so should be ok as well.

EDIT: the gdiplus plugin was changed a bit codewise as the autotool based makefile.am seems to have been for a very old mingw.org build using the library from the windows sdk because it was missing in the mingw.org windows api at the time. but both mingw-w64 and mingw.org now has the gdiplus sdk integrated so it seems it is no longer nessesary with this hack.

**Reply by revelator** [reply]
In fact i could remove the whole patch shenanigans for gdiplus from the makefile as it was only intended to coerce mingw to build it and the patches are missing anyway, they are however located here https://gitlab.com/graphviz/graphviz/-/tree/d60f51cb93557f08cd415ee02f4abe71d7bac3f8/plugin/gdiplus and the cmake build does not seem to use them anymore either.

**Reply by revelator** [reply]
Well i kinda managed to preserve the swig code so that it should work on nix also but it got really messy and actually made the patch even bigger :S since i use if WITH_WIN32 else endif blocks in the makefile.
There has to be a better way like this maybe ->

if WITH_WIN32
SO_DLL=dll
else
SO_DLL=so
endif

and then do something like this maybe -> cp -f $(DESTDIR)$(pkgpythondir)/libgv_python.$(SO_DLL) _gv.$(SO_DLL); ?

**Reply by revelator** [reply]
[013-further-fixes.zip](https://github.com/msys2/MINGW-packages/files/6112243/013-further-fixes.zip)

well here we go is this more to your taste :)

**Reply by revelator** [reply]
well turns out --enable-python=yes enabes the old way of linking with the python interpreter the downside is that it cannot distinguish between python2 and python3 so it just links to the first one found. the swig interface was created to work around this in an easier way but enabling both swig and the above yields 3 python interfaces :facepalm: i have a sinking feeling much of this code has been left in to keep compatibility with some really old compilers and just forgotten in the long run :S. A thorough spring cleaning might be in order soon.

**Comment by lazka** [comment]
I guess we should also try to catch up with upstream releases some time (2.47.0  will be out next week it seems), and maybe try to move to cmake, since that is what upstream is focusing on right now.

**Reply by revelator** [reply]
aye that would be nice :) though atm the cmake based build seems to focus on msvc on windows so some patching in that regard would be required.

**Reply by revelator** [reply]
btw a smal snag i just noticed is that gvedit does not work correctly if executed from bash since it looks for it's configuration files in ../share/graphviz a user would first need to chdir to /mingw*/bin or else it will croak about not finding attrs.txt. if executed directly it works fine though.

**Reply by revelator** [reply]
could maybe create a bash shell script wrapper for this so that users who want to use it that way would be able to.

**Reply by revelator** [reply]
`#!/bin/bash
basedir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd -W )"
echo "starting gvedit from $basedir"
pushd "$basedir" >/dev/null
"$basedir/gvedit.exe" "$@"
popd >/dev/null`

works as a charm :)

**Reply by revelator** [reply]
hehe can even direct it to a dir containing graphs and have it open it directly with this nice 👍
